openedx/adg/lms/applications/tests/test_views.py

```diff
@@ -0,0 +1,162 @@
+        self.assertFalse(user_application.is_written_application_submitted)
+        self.assertTrue(user_application.is_prerequisite_courses_passed)
+        self.assertEqual(self.response.context_data['percentage'], "50%")
+
+    def test_complete_application_done_state(self):
+        """
+        Test the case where the user has completed both objectives i.e the pre_req courses and the
+        written application
+        """
+        self.set_application_state("is_prerequisite_courses_passed", "is_written_application_submitted"),
+
+        user_application = self.response.context_data['user_application']
+        self.assertTrue(user_application.is_written_application_submitted)
+        self.assertTrue(user_application.is_prerequisite_courses_passed)
+        self.assertEqual(self.response.context_data['percentage'], "100%")
+
+
+class ApplicationSubmitViewTest(TestCase):
+    """
+    TestCase for ApplicationSubmitView
+    """
+
+    def setUp(self):
+        """
+        Setup the user and the user application for testing.
+        """
+        self.user = UserFactory(username=USERNAME, email=EMAIL, password=PASSWORD)
+        self.user_application = ApplicationHub.objects.create(user=self.user)
+        self.user_application.is_prerequisite_courses_passed = True
+        self.user_application.is_written_application_submitted = True
+        self.user_application.is_application_submitted = False
+        self.user_application.save()
+        self.client = Client()
+
+    def test_if_redirects_without_login(self):
+        """
+        Test the case where an unauthenticated user is redirected to login page or not.
+        """
+        response = self.client.get(reverse('application_submit'))
+        self.assertRedirects(response, '/login?next=/application/submit')
+
+    def test_logged_in_user_without_required_objectives(self):
+        """
+        Test the case where an authenticated user hits the url without having completed the required objectives i.e
+        the pre_req courses and the written application.
+        """
+        self.user_application.is_prerequisite_courses_passed = False
+        self.user_application.is_written_application_submitted = False
+
+        self.client.login(username=USERNAME, password=PASSWORD)
+        self.user_application.save()
+        response = self.client.get(reverse('application_submit'))
+        self.assertEqual(response.status_code, 400)
+
+    @mock.patch('openedx.adg.lms.applications.views.send_mandrill_email')
+    def test_logged_in_user(self, send_mandrill_mail):
+        """
+        Test the case where an authenticated user, with all the required objectives completed, hits the url.
+        """
+        self.client.login(username=USERNAME, password=PASSWORD)
+        response = self.client.get(reverse('application_submit'))
+        self.assertEqual(response['Location'], reverse('application_success'))
+        self.assertEqual(response.status_code, 302)
```

Comment:
Like I've explained in the above comment, I had to avoid the template middleware (as the setup cannot find the templates anyway) I used the `RequestFactory`. In the case of this view, no template is associated with this, so we may use `Client` without having to worry about the template directory not being found.